### PR TITLE
Issue 2430 - avoid workload usage if not needed

### PR DESCRIPTION
**Pull request by dlarson04**
- If multiple versions are in a policy, check compatibility of constraints in DP first
- Don't use workload_usage in DB unless needed
- Move getting maxHeartbeat until before we need to send a message

### CODE REVIEW

File: agreementbot/agreementworker.go [hunk 1]
```diff
@@ -331,15 +331,73 @@ func (b *BaseAgreementWorker) AgreementLockManager() *AgreementLockManager {
...
+                // non pattern case
+
+                if err != nil {
+                        glog.Errorf(BAWlogstring(workerId, fmt.Sprintf("%v", err)))
+                        return
+                } else if nodePolicy == nil {
+                        glog.Warning(BAWlogstring(workerId, fmt.Sprintf("Cannot find node policy for this node %v.", wi.Device.Id)))
+                        return
+                } else {
+                        if glog.V(5) {glog.Infof(BAWlogstring(workerId, fmt.Sprintf("retrieved node policy: %v", nodePolicy)))}
```
Comment (dabooz) [inline agreementbot/agreementworker.go:369]: just use glog.V(5).Infof

File: agreementbot/agreementworker.go [hunk 2]
```diff
@@ -331,15 +331,73 @@ func (b *BaseAgreementWorker) AgreementLockManager() *AgreementLockManager {
...
+        }
+
 	// Generate an agreement ID
 	agreementIdString, aerr := cutil.GenerateAgreementId()
 	if aerr != nil {
 		glog.Errorf(BAWlogstring(workerId, fmt.Sprintf("error generating agreement id %v", aerr)))
 		return
 	}
-	glog.V(5).Infof(BAWlogstring(workerId, fmt.Sprintf("using AgreementId %v", agreementIdString)))
+	if glog.V(5) {glog.Infof(BAWlogstring(workerId, fmt.Sprintf("using AgreementId %v", agreementIdString)))}
```
Comment (dabooz) [inline agreementbot/agreementworker.go:400]: Just use glog.V(5).Infof

File: agreementbot/agreementworker.go [hunk 3]
```diff
@@ -368,35 +426,30 @@ func (b *BaseAgreementWorker) InitiateNewAgreement(cph ConsumerProtocolHandler,
...
 	// each workload and whether or not this workload has been tried before. Also, iterate the loop more than once if we choose
 	// a workload entry that turns out to be unsupportable by the device.
-	msgPrinter := i18n.GetMessagePrinter()
 	foundWorkload := false
 	var workload, lastWorkload *policy.Workload
 	svcIds := []string{} // stores the service ids for all the services, top level and dependent services
 	found := true        // if the service policy can be found from the businesspol_manager
 	var servicePol *externalpolicy.ExternalPolicy
 
+        var wlUsage *persistence.WorkloadUsage = nil
```
Comment (dabooz) [inline agreementbot/agreementworker.go:440]: some kind of a white space/tab thing going on here. If you run make format, it might clean this up for you.

**Review by dabooz — CHANGES_REQUESTED** [review]

File: agreementbot/agreementworker.go [hunk 4]
```diff
@@ -685,12 +723,24 @@ func (b *BaseAgreementWorker) InitiateNewAgreement(cph ConsumerProtocolHandler,
 				workload.DeploymentSignature = workloadDetails.GetDeploymentSignature()
 			}
 
-			glog.V(5).Infof(BAWlogstring(workerId, fmt.Sprintf("workload %v is supported by device %v", workload, wi.Device.Id)))
+			if glog.V(5) {glog.Infof(BAWlogstring(workerId, fmt.Sprintf("workload %v is supported by device %v", workload, wi.Device.Id)))}
```
[the remark this answers is not in the document]
Comment (dlarson04) [reply agreementbot/agreementworker.go:726]: talked to Dave... its an optimization.. will leave as added

File: agreementbot/agreementworker.go (hunk 1 above)
Comment (dlarson04) [reply agreementbot/agreementworker.go:369]: talked to Dave... its an optimization.. will leave as added

File: agreementbot/agreementworker.go (hunk 2 above)
Comment (dlarson04) [reply agreementbot/agreementworker.go:400]: talked to Dave... its an optimization.. will leave as added

File: agreementbot/agreementworker.go [hunk 5]
```diff
@@ -331,15 +331,73 @@ func (b *BaseAgreementWorker) AgreementLockManager() *AgreementLockManager {
...
+        }
+}
+
 func (b *BaseAgreementWorker) InitiateNewAgreement(cph ConsumerProtocolHandler, wi *InitiateAgreement, random *rand.Rand, workerId string) {
 
+        msgPrinter := i18n.GetMessagePrinter()
+
+        // get node policy
+        nodePolicyHandler := exchange.GetHTTPNodePolicyHandler(b)
+        _, nodePolicy, err := compcheck.GetNodePolicy(nodePolicyHandler, wi.Device.Id, msgPrinter)
```
[the remark this answers is not in the document]
Comment (dlarson04) [reply agreementbot/agreementworker.go:356]: good catch... done

File: agreementbot/agreementworker.go (hunk 3 above)
Comment (dlarson04) [reply agreementbot/agreementworker.go:440]: done

**Review by dabooz — APPROVED** [review]

**Pull request merged**